Fix various issues with build-git-installers.yml

mjcheetham opened the pull request:
There are several issues that have been uncovered with the changes made in #738. Let's fix them!

* Check out `akv-secret` action before it is used.
* Log in to Azure before accessing the Key Vault.
* Don't mask empty lines.
* Use a buffer to fix encoding issues when writing binary data.
* Correctly mask multi-line secret values.
* Add missing `require('path')` statement.

dscho committed:
fixup! .github/actions/akv-secret: add action to get secrets

This fixes "Error: path is not defined" issues in the workflow run.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fixup! release: create initial Windows installer build workflow

Since the Azure Portal supports only single-line secrets (but those
lines can be _very_ long), the convention is to store GPG keys (which
_are_ multi-line) as base64-encoded values.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fixup! release: add Mac OSX installer build

The custom Action is checked out at a different location than
`.github/`...

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fixup! release: add signing step for .deb package

The `akv-secret` Action must be checked out before it can be used.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fixup! build-git-installers: publish gpg public key

The `akv-secret` Action must be checked out before it can be used.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fixup! release: create initial Windows installer build workflow

We need to log into Azure, and check out the custom Action, before we
can download secrets from the KeyVault via `akv-secret`.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! fixup! release: add Mac OSX installer build

We need to log into Azure before we can access the KeyVault.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! .github/actions/akv-secret: add action to get secrets

Use a buffer rather than a string when handling the output from the `az`
command. Handling of binary data (that was base64 encoded) requires that
we use a buffer and not a string, or else we will end up writing invalid
data to files/output variables.

Introduce several new helper functions for working with buffers,
including trimming the EOL bytes (CR, LF), and fix up some of the
output functions to correctly validate the value passed - it must be
something printable (UTF-8-ish).

Finally ensure that we correctly mask multi-line secret values by
emitting a `::add-mask` command for each line.

Co-authored-by: Johannes Schindelin <johannes.schindelin@gmx.de>
Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! fixup! release: add Mac OSX installer build

dscho committed:
fixup! fixup! release: add Mac OSX installer build

We also need to quote the Apple App identity (because it contains a
parenthesis).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

mjcheetham committed:
fixup! fixup! release: create initial Windows installer build workflow

Ensure that we don't clobber existing directory with the checkout,
which in this instance contains downloaded artifacts!

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! fixup! release: add Mac OSX installer build

We're not using quotes around simple string values like 'git' in other
places.. let's be consistent.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! fixup! release: add signing step for .deb package

Ensure that we don't clobber existing directory with the checkout,
which in this instance contains downloaded artifacts!

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! fixup! build-git-installers: publish gpg public key

Ensure that we don't clobber existing directory with the checkout,
which in this instance contains downloaded artifacts!

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! .github/actions/akv-secret: add action to get secrets

Don't mask empty lines since this isn't possible!

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! fixup! release: create initial Windows installer build workflow

Use the global `DO_WIN_CODESIGN` environment variable rather than
re-compute this for the verification of codesigning step.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
fixup! fixup! release: create initial Windows installer build workflow

Replace Bash script (using `signtool`) for validating executables are
code-signed correctly with a PowerShell script (which instead uses the
`Get-AuthenticodeSignature` cmdlet).

The `signtool` is only available in the Windows SDK, which isn't always
installed on self-hosted runners (e.g., for ARM64), but PowerShell is
always available on our images.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

derrickstolee commented on .github/actions/akv-secret/index.js:
```diff
@@ -1,6 +1,8 @@
 const fs = require('fs');
 const os = require('os');
+const path = require('path');
+const { isUtf8 } = require("buffer")
```
Do you need a semicolon here?

dscho replied:
In JavaScript, most semicolons are actually not needed. In this instance, while I agree that it would be more consistent with the remainder of the code's style to have a semicolon, I am loathe to deviate from the tagged v2.49.0.vfs.0.1... Maybe as a `fixup!` on top of e730d37255bba38d14fb909a750a37aa4bd30d91?

mjcheetham replied:
Pushed on top.

mjcheetham committed:
fixup! .github/actions/akv-secret: add action to get secrets

Add semi-colon to the `require('buffer')` line to stay consistent with
the rest of the index.js file.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

dscho reviewed (approved):
The changes all look good to me!
